funsor/sum_product.py

```diff
@@ -45,6 +45,67 @@ def _partition(terms, sum_vars):
     return components
 
 
+def _unroll_plate(factors, plate, var_to_ordinal):
+    # replicated variables
+    plate_vars = set()
+    for var, ordinal in var_to_ordinal.items():
+        if plate in ordinal:
+            plate_vars.add(var)
+
+    # unroll factors
+    unrolled_factors = []
+    for factor in factors:
+        if plate in factor.inputs:
+            size = factor.inputs[plate].size
+            f_vars = plate_vars.intersection(factor.inputs)
+            unrolled_factors.extend([factor(
+                    **{plate: i},
+                    **{var: "{}_{}{}".format(var, plate, i) for var in f_vars}
+                ) for i in range(size)])
+        else:
+            unrolled_factors.append(factor)
+
+    # unroll variables
+    for var in plate_vars:
+        ordinal = var_to_ordinal.pop(var)
+        new_ordinal = ordinal.difference(plate)
+        var_to_ordinal.update({"{}_{}{}".format(var, plate, i): new_ordinal
+                               for i in range(size)})
```

Comment:
`size` on this line is defined in the for loop above. There is probably a better way to do this ...

Reply:
How about `size = next([f.inputs[plate].size for f in factors if plate in f.inputs])`?